client/build_test.go

```diff
@@ -175,3 +189,554 @@ func testUnknownBuildID(t *testing.T, sb integration.Sandbox) {
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerMounts is testing mounts derived from various
+// llb.States
+func testClientGatewayContainerMounts(t *testing.T, sb integration.Sandbox) {
+	if sb.Rootless() {
+		// TODO fix this
+		// We get `panic: cannot statfs cgroup root` when running this test
+		// with runc-rootless
+		t.Skip("Skipping runc-rootless for cgroup error")
+	}
+	requiresLinux(t)
+
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	tmpdir, err := ioutil.TempDir("", "buildkit-buildctl")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	err = ioutil.WriteFile(filepath.Join(tmpdir, "local-file"), []byte("local"), 0644)
+	require.NoError(t, err)
+
+	a := agent.NewKeyring()
+	sockPath, clean, err := makeSSHAgentSock(a)
+	require.NoError(t, err)
+	defer clean()
+
+	ssh, err := sshprovider.NewSSHAgentProvider([]sshprovider.AgentConfig{{
+		ID:    t.Name(),
+		Paths: []string{sockPath},
+	}})
+	require.NoError(t, err)
+
+	product := "buildkit_test"
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		mounts := map[string]llb.State{
+			"/": llb.Image("busybox:latest").Run(
+				llb.Shlex("touch /root-file /cached/cache-file"),
+				llb.AddMount("/cached", llb.Scratch(), llb.AsPersistentCacheDir(t.Name(), llb.CacheMountShared)),
+			).Root(),
+			"/foo": llb.Image("busybox:latest").Run(
+				llb.Shlex("touch foo-file"),
+				llb.Dir("/tmp"),
+				llb.AddMount("/tmp", llb.Scratch()),
+			).GetMount("/tmp"),
+			"/local": llb.Local("mylocal"),
+			// TODO How do we get a results.Ref for a cache mount, tmpfs mount
+		}
+
+		containerMounts := []client.Mount{{
+			Dest:      "/cached",
+			MountType: pb.MountType_CACHE,
+			CacheOpt: &pb.CacheOpt{
+				ID:      t.Name(),
+				Sharing: pb.CacheSharingOpt_SHARED,
+			},
+		}, {
+			Dest:      "/tmpfs",
+			MountType: pb.MountType_TMPFS,
+		}, {
+			Dest:      "/run/secrets/mysecret",
+			MountType: pb.MountType_SECRET,
+			SecretOpt: &pb.SecretOpt{
+				ID: "/run/secrets/mysecret",
+			},
+		}, {
+			Dest:      sockPath,
+			MountType: pb.MountType_SSH,
+			SSHOpt: &pb.SSHOpt{
+				ID: t.Name(),
+			},
+		}}
+
+		for mountpoint, st := range mounts {
+			def, err := st.Marshal(ctx)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to marshal state")
+			}
+
+			r, err := c.Solve(ctx, client.SolveRequest{
+				Definition: def.ToPB(),
+			})
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to solve")
+			}
+			containerMounts = append(containerMounts, client.Mount{
+				Dest:      mountpoint,
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			})
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{Mounts: containerMounts})
+		if err != nil {
+			return nil, err
+		}
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args:   []string{"sleep", "10"},
+			Cwd:    "/",
```

Comment:
Currently it is a required field, otherwise runc fails with `Cwd property must not be empty`. I could not figure out where there is defaulted on the `ExecOp` solve side of things, there is no default in the executors, maybe there should be?

Reply:
@hinshun helped me track down where the cwd was defaulted on the llb/solve side:
https://github.com/moby/buildkit/blob/master/client/llb/exec.go#L167

I have updated to include default Cwd in the container Start requests: https://github.com/moby/buildkit/pull/1627/commits/2f0738c47aea572a0d6abe96c10aba3d910a9643